README.md
# winston-syslog

A Syslog transport for [winston][0].

[![Version npm](https://img.shields.io/npm/v/winston-syslog.svg?style=flat-square)](https://www.npmjs.com/package/winston-syslog)
[![npm Downloads](https://img.shields.io/npm/dm/winston-syslog.svg?style=flat-square)](https://npmcharts.com/compare/winston-syslog?minimal=true)
[![build status](https://github.com/winstonjs/winston-syslog/actions/workflows/ci.yml/badge.svg)](https://github.com/winstonjs/winston-syslog/actions/workflows/ci.yml)

[![NPM](https://nodei.co/npm/winston-syslog.png?downloads=true&downloadRank=true)](https://nodei.co/npm/winston-syslog/)

## Requirements

* winston >= 3.0.0

## Installation

### Installing npm (node package manager)

``` bash
  $ curl http://npmjs.org/install.sh | sh
```

### Installing winston-syslog

``` bash
  $ npm install winston
  $ npm install winston-syslog
```

## Motivation
`tldr;?`: To break the [winston][0] codebase into small modules that work together.

The [winston][0] codebase has been growing significantly with contributions and other logging transports. This is **awesome**. However, taking a ton of additional dependencies just to do something simple like logging to the Console and a File is overkill.

## Usage
To use the Syslog transport in [winston][0], you simply need to require it and then either add it to an existing [winston][0] logger or pass an instance to a new [winston][0] logger:

``` js
  const winston = require('winston');

  //
  // Requiring `winston-syslog` will expose
  // `winston.transports.Syslog`
  //
  require('winston-syslog').Syslog;

  winston.add(new winston.transports.Syslog(options));
```

In addition to the options accepted by the syslog (compliant with [RFC 3164][1] and [RFC 5424][2]), the Riak transport also accepts the following options. It is worth noting that the riak-js debug option is set to *false* by default:

* __host:__ The host running syslogd, defaults to localhost.
* __port:__ The port on the host that syslog is running on, defaults to syslogd's default port.
* __protocol:__ The network protocol to log over (e.g. `tcp4`, `udp4`, `tls4`, `unix`, `unix-connect`, etc), defaults to `udp4`.
* __protocolOptions:__ Socket connect options. See [`net.socket.connect`](https://nodejs.org/api/net.html#net_socket_connect_options_connectlistener) for available options.
* __path:__ The path to the syslog dgram socket (i.e. `/dev/log` or `/var/run/syslog` for OS X).
* __pid:__ PID of the process that log messages are coming from (Default `process.pid`).
* __facility:__ Syslog facility to use (Default: `local0`).
* __localhost:__ Host to indicate that log messages are coming from (Default: `localhost`).
* __type:__ The type of the syslog protocol to use (Default: `BSD`, also valid: `'3164'`, `'5424'`, `'RFC3164'` or `'RFC5424'`).
* __app_name:__ The name of the application (Default: `process.title`).
* __eol:__ The end of line character to be added to the end of the message
(Default: Message without modifications).
* __secureProtocol:__ See <https://nodejs.org/api/tls.html#tlscreatesecurecontextoptions> for more information on this option,
passed through from [this](https://nodejs.org/api/tls.html#new-tlstlssocketsocket-options) constructor.
* __ciphers:__ See <https://nodejs.org/api/tls.html#tlscreatesecurecontextoptions> for more information on this option,
passed through from [this](https://nodejs.org/api/tls.html#new-tlstlssocketsocket-options) constructor.
* __ecdhCurve:__ See <https://nodejs.org/api/tls.html#tlscreatesecurecontextoptions> for more information on this option,
passed through from [this](https://nodejs.org/api/tls.html#new-tlstlssocketsocket-options) constructor.

*Metadata:* Logged as string compiled by [glossy][3].

By default, syslog messages are produced by [glossy][3], but you can override that behavior by providing a
custom **Producer** instance via the **customProducer** setting.

## Log Levels
Because syslog only allows a subset of the levels available in [winston][0], levels that do not match will be ignored. Therefore, in order to use `winston-syslog` effectively, you should indicate to [winston][0] that you want to use the syslog levels:

``` js
  const winston = require('winston');
  const logger = winston.createLogger({
    levels: winston.config.syslog.levels,
    transports: [
      new winston.transports.Syslog()
    ]
  });
```

The `Syslog` transport will only log to the level that are available in the syslog protocol. These are (in increasing order of severity):

* debug
* info
* notice
* warning
* err
* crit
* alert
* emerg

## Syslog Configuration

You will have to configure your syslog server to accept TCP or UDP connections.
This is usually done in `/etc/syslog-ng.conf`. Let's say you have an app called `fnord`
and want to use TCP, the configuration would look something like this:

```
  source tcp_s {
    tcp(ip(0.0.0.0) port(514) max-connections(256));
  };
  destination fnord_d {
    file("/var/log/fnord.log");
  };
  log { source(tcp_s); destination(fnord_d); };
```

If you have multiple apps which need to log via TCP, you can specify filters, as such:

```
  filter fnord_f { program("fnord"); };
```

Then modify the log statement to read:

```
  log { source(tcp_s); filter(fnord_f); destination(fnord_d); };
```

Now if you have another app, called `bnord`, create similar `destination` and `filter` configurations for it, and specify a new log statement, with the same `source`:

```
  log { source(tcp_s); filter(bnord_f); destination(bnord_d); };
```

For this to work, you have to make sure you set the `process.title` variable in your node app.

``` js
  process.title = 'fnord';
```

#### Author: [Charlie Robbins](http://blog.nodejitsu.com)
#### Contributors: [Squeeks](https://github.com/squeeks)

[0]: https://github.com/indexzero/winston
[1]: http://www.ietf.org/rfc/rfc3164.txt
[2]: http://tools.ietf.org/html/rfc5424
[3]: https://github.com/squeeks/glossy


CHANGELOG.md
# CHANGELOG

## v2.6.0 / 2022-06-28

- (@wbt) Bump dependencies
- #[192], (@Omidznlp) Add TLSsocket Options

## v2.5.0 / 2022-02-04

- #[165], (@Apollon77) Catch exceptions thrown by socket send
- (@DABH) Bump dependencies

## v2.4.0 / 2020-01-01

- (@DABH) Node v12 support, fix node-unix-dgram issues, update dependencies
- #[115], (@pepakriz) TLS connection support
- #[123], (@JeffTomlinson, @elliotttf)  Handle oversize messages sent over UDP transports
- #[116], (@pepakriz) Make socket options configurable
- #[122], (@cjbarth) Correct improper opening and closing of sockets

## v2.2.0 / 2019-08-14

- #[82], (@AlexMost) allow use of a customer producer
- #[109], (@gdvyas) Prevent error before connection is established
- #[114], (@vrza) Support 'udp' as an alias of 'udp4'

## v2.1.0 / 2019-02-17

- (@DABH) Fix tests by fixing error emission/handling
- #[108], (@vrza) Make winston 3 a peer dependency
- #[102], (@stieg) Require winston >= 3 and add corresopnding note in readme
- #[105], (@mohd-akram) Update dependencies for latest Node compatibility



test/tls-test.js
const vows = require('vows');
const assert = require('assert');
const selfsigned = require('selfsigned');
const tls = require('tls');
const winston = require('winston');
require('../lib/winston-syslog').Syslog;

// ----- Constants
const HOST = 'localhost';
const PORT = 10514;
const PROMISE_TIMEOUT = 1000;

// ----- Helpers
function wrapToPromise(target, event) {
  return new Promise((resolve, reject) => {
    const timeout = setTimeout(() => {
      reject('Timeout for event promise');
    }, PROMISE_TIMEOUT);
    target.on(event, (...args) => {
      clearTimeout(timeout);
      resolve(...args);
    });
  });
}

function nodeMajorVersion() {
  return Number.parseInt(process.version.match(/^v(\d+\.\d+)/)[1], 10);
}

// ----- Certificate handling
function generateCertificate() {
  // Generate server and client certificates
  const attributes = [{ name: 'commonName', value: 'localhost' }];
  const x509 = selfsigned.generate(attributes, {
    days: 1,
    clientCertificate: true,
    extensions: [
      {
        name: 'keyUsage',
        keyCertSign: true,
        digitalSignature: true,
        nonRepudiation: true,
        keyEncipherment: true,
        dataEncipherment: true
      },
      {
        name: 'subjectAltName',
        altNames: [
          {
            type: 2, // DNS
            value: 'localhost'
          }
        ]
      }
    ]
  });
  return x509;
}

// ----- TLS configs
const x509 = generateCertificate();
const validServerTLS = {
  key: x509.private,
  cert: x509.cert,
  ca: [x509.cert],
  rejectUnauthorized: true,
  requestCert: true
};

const validClientTLS = {
  key: x509.clientprivate,
  cert: x509.clientcert,
  ca: [x509.cert],
  rejectUnauthorized: true
};

const untrustedClientTLS = {
  ...validClientTLS,
  ca: null
};

const missingClientTLS = {
  ca: null,
  key: null,
  cert: null,
  rejectUnauthorized: false
};

// ----- TLS Server
const serverEvents = {
  data: 'data',
  tlsClientError: 'tlsClientError',
  error: 'error',
  listening: 'listening',
  secureConnection: 'secureConnection'
};

async function startServer({ host = HOST, port = PORT, tlsOpts } = {}) {
  const server = tls.createServer({ ...tlsOpts });
  let clients = [];
  server.on('secureConnection', (client) => {
    clients.push(client);
    client.on('close', () => {
      clients = clients.filter((c) => c !== client);
    });
    client.on('data', (data) => {
      server.emit(serverEvents.data, data.toString());
    });
  });
  server.forceClose = () => {
    clients.forEach((client) => client.destroy());
    server.close();
  };
  server.listen({ host, port });
  await wrapToPromise(server, serverEvents.listening);
  return server;
}

// ----- Init Logger
function initLogger({ host = HOST, port = PORT, tlsOpts } = {}) {
  const syslogOptions = {
    host,
    port,
    protocol: 'tls4',
    protocolOptions: { ...tlsOpts }
  };
  const logger = winston.createLogger({
    transports: [new winston.transports.Syslog(syslogOptions)]
  });
  return logger;
}

// ----- Test Cases
const TEST_MESSAGE = 'Test Message';
const SYSLOG_FORMAT = `"message":"${TEST_MESSAGE}"`;

let serverInstance;

vows
  .describe('tls-connect')
  .addBatch({
    'Trying to connect to a TLS server with mutual TLS': {
      'topic': function () {
        startServer({ tlsOpts: validServerTLS }).then((server) => {
          serverInstance = server;
          const promise = wrapToPromise(server, serverEvents.data);
          initLogger({ tlsOpts: validClientTLS }).info(TEST_MESSAGE);
          promise.then(msg => this.callback(null, msg));
        });
      },
      'TLS server should receive log message': function (msg) {
        assert.include(msg, SYSLOG_FORMAT);
      },
      'teardown': function () {
        serverInstance.forceClose();
      }
    }
  })
  .addBatch({
    'Trying to connect to a TLS server with untrusted certificate': {
      'topic': function () {
        startServer({ tlsOpts: validServerTLS }).then((server) => {
          serverInstance = server;
          const logger = initLogger({ tlsOpts: untrustedClientTLS });
          logger.on('error', (loggerError) => {
            this.callback(null, loggerError);
          });
          logger.info(TEST_MESSAGE);
        });
      },
      'Client should refuse connection': function (e, loggerError) {
        assert.strictEqual(loggerError.code, 'DEPTH_ZERO_SELF_SIGNED_CERT');
        assert.include(loggerError.message, 'self signed certificate');
      },
      'teardown': function () {
        serverInstance.forceClose();
      }
    }
  })
  .addBatch({
    'Trying to connect to a TLS server without client certificate': {
      'topic': function () {
        startServer({ tlsOpts: validServerTLS }).then((server) => {
          serverInstance = server;
          const promise = wrapToPromise(server, serverEvents.tlsClientError);
          const logger = initLogger({ tlsOpts: missingClientTLS });
          logger.on('error', (loggerError) => {
            promise
              .then((serverError) => this.callback(null, loggerError, serverError))
              .catch((error) => this.callback(error, loggerError, null));
          });
          logger.info(TEST_MESSAGE);
        });
      },
      'Server should refuse connection': function (e, loggerError, serverError) {
        // Client and Server error type changes between Node versions
        if (nodeMajorVersion() >= 12) {
          assert.strictEqual(loggerError.code, 'ERR_SSL_TLSV13_ALERT_CERTIFICATE_REQUIRED');
          assert.include(loggerError.message, 'alert number 116');
          assert.strictEqual(serverError.code, 'ERR_SSL_PEER_DID_NOT_RETURN_A_CERTIFICATE');
          assert.include(serverError.message, 'peer did not return a certificate');
        } else {
          assert.strictEqual(loggerError.code, 'EPROTO');
          assert.include(loggerError.message, 'alert number 40');
          assert.include(serverError.message, 'peer did not return a certificate');
        }
      },
      'teardown': function () {
        serverInstance.forceClose();
      }
    }
  })
  .export(module);


test/format-test.js
'use strict';

const vows = require('vows');
const assert = require('assert');
const Syslog = require('../lib/winston-syslog.js').Syslog;
const dgram = require('dgram');
const parser = require('@myndzi/glossy').Parse;

const PORT = 11229;
let server;
let transport;

const { MESSAGE, LEVEL } = require('triple-beam');

vows
  .describe('syslog messages')
  .addBatch({
    'opening fake syslog server': {
      'topic': function () {
        const self = this;
        server = dgram.createSocket('udp4');
        server.on('listening', function () {
          self.callback();
        });

        server.bind(PORT);
      },
      'default format': {
        'topic': function () {
          const self = this;
          server.once('message', function (msg) {
            parser.parse(msg, function (d) {
              self.callback(null, d);
            });
          });

          transport = new Syslog({
            port: PORT
          });
          transport.log({ [LEVEL]: 'debug', [MESSAGE]: 'ping' }, function (err) {
            assert.ifError(err);
          });
        },
        'should have host field set to localhost': function (msg) {
          assert.equal(msg.host, 'localhost');
          transport.close();
        },
        'setting locahost option to a different falsy value (null)': {
          'topic': function () {
            const self = this;
            server.once('message', function (msg) {
              parser.parse(msg, function (d) {
                self.callback(null, d);
              });
            });

            transport = new Syslog({
              port: PORT,
              localhost: null
            });

            transport.log({ [LEVEL]: 'debug', [MESSAGE]: 'ping2' }, function (
              err
            ) {
              assert.ifError(err);
            });
          },
          'should have host different from localhost': function (msg) {
            assert.notEqual(msg.host, 'localhost');
            transport.close();
          },
          'setting appName option to hello and type to 5424': {
            'topic': function () {
              const self = this;
              server.once('message', function (msg) {
                parser.parse(msg, function (d) {
                  self.callback(null, d);
                });
              });

              transport = new Syslog({
                port: PORT,
                type: '5424',
                appName: 'hello'
              });

              transport.log(
                { [LEVEL]: 'debug', [MESSAGE]: 'app name test' },
                function (err) {
                  assert.ifError(err);
                }
              );
            },
            'should have appName field set to hello': function (msg) {
              assert.equal(msg.appName, 'hello');
            },
            'should have proper time zone': function (msg) {
              const now = new Date();
              assert.equal((Math.abs((now - msg.time) / 1000) < 300), true);
            },
            'should have proper milliseconds format': function (msg) {
              const milliseconds = msg.originalMessage.split(' ')[1].slice(20, 23);
              assert.equal(isNaN(milliseconds), false);
              transport.close();
            },
            'setting app_name option to hello': {
              'topic': function () {
                const self = this;
                server.once('message', function (msg) {
                  parser.parse(msg, function (d) {
                    self.callback(null, d);
                  });
                });

                transport = new Syslog({
                  port: PORT,
                  type: '5424',
                  app_name: 'hello'
                });

                transport.log(
                  { [LEVEL]: 'debug', [MESSAGE]: 'app name test' },
                  function (err) {
                    assert.ifError(err);
                  }
                );
              },
              'should have appName field set to hello': function (msg) {
                assert.equal(msg.appName, 'hello');
                transport.close();
              }
            }
          }
        }
      },
      'teardown': function () {
        server.close();
      }
    }
  })
  .addBatch({
    'opening fake syslog server': {
      'topic': function () {
        var self = this;
        server = dgram.createSocket('udp4');
        server.on('listening', function () {
          self.callback();
        });

        server.bind(PORT);
      },
      'Custom producer': {
        'topic': function () {
          var self = this;
          server.once('message', function (msg) {
            self.callback(null, msg.toString());
          });

          function CustomProducer() {}
          CustomProducer.prototype.produce = function (opts) {
            return 'test ' + opts.severity + ': ' + opts.message;
          };

          transport = new Syslog({
            port: PORT,
            customProducer: CustomProducer
          });

          transport.log({ [LEVEL]: 'debug', [MESSAGE]: 'ping' }, function (err) {
            assert.ifError(err);
          });
        },
        'should apply custom syslog format': function (msg) {
          assert.equal(msg, 'test debug: ping');
          transport.close();
        }
      },
      'teardown': function () {
        server.close();
      }
    }
  })
  .export(module);


test/syslog-test.js
/* eslint new-cap: ["error", { "newIsCapExceptions": ["createLogger"] }] */
/*
 * syslog-test.js: Tests for instances of the Syslog transport
 *
 * (C) 2010 Charlie Robbins
 * MIT LICENSE
 *
 */

const vows = require('vows');
const assert = require('assert');
const winston = require('winston');
const Syslog = require('../lib/winston-syslog').Syslog;

function assertSyslog(transport) {
  assert.instanceOf(transport, Syslog);
  assert.isFunction(transport.log);
  assert.isFunction(transport.connect);
}

function closeTopicInfo() {
  const transport = new winston.transports.Syslog();
  const logger = new winston.createLogger({ transports: [transport] });

  logger.log('info', 'Test message to actually use socket');
  logger.remove(transport);

  return transport;
}

function closeTopicDebug() {
  const transport = new winston.transports.Syslog();
  const logger = new winston.createLogger({ transports: [transport] });

  logger.log('debug', 'Test message to actually use socket');
  logger.remove(transport);

  return transport;
}

var transport = new Syslog();

vows.describe('winston-syslog').addBatch({
  'An instance of the Syslog Transport': {
    'should have the proper methods defined': function () {
      assertSyslog(transport);
    },
    'teardown': function () {
      transport.close();
    },
    'on close after not really writing': {
      topic: closeTopicDebug,
      on: {
        closed: {
          'closes the socket': function (socket) {
            assert.isNull(socket);
          }
        }
      }
    },
    'on close after really writing': {
      topic: closeTopicInfo,
      on: {
        closed: {
          'closes the socket': function (socket) {
            assert.isNull(socket._handle);
          }
        }
      }
    },
    'localhost option': {
      'should default to localhost': function () {
        const transportLocal = new winston.transports.Syslog();
        assert.equal(transportLocal.localhost, 'localhost');
        transportLocal.close();
      },
      'should accept other falsy entries as valid': function () {
        let transportNotLocal = new winston.transports.Syslog({ localhost: null });
        assert.isNull(transportNotLocal.localhost);
        transportNotLocal.close();
        transportNotLocal = new winston.transports.Syslog({ localhost: false });
        assert.equal(transportNotLocal.localhost, false);
        transportNotLocal.close();
      }
    },
    'adding / removing transport to syslog': {
      'should just work': function () {
        winston.add(new winston.transports.Syslog());
        winston.remove(new winston.transports.Syslog());
        winston.add(new winston.transports.Syslog());
        winston.remove(new winston.transports.Syslog());
      }
    }
  }
}).export(module);


lib/winston-syslog.js
/*
 * syslog.js: Transport for logging to a remote syslog consumer
 *
 * (C) 2011 Squeeks and Charlie Robbins
 * MIT LICENCE
 *
 */

const dgram = require('dgram');
const net = require('net');
const secNet = require('tls');
const utils = require('./utils');
const glossy = require('glossy');
const winston = require('winston');
const Transport = require('winston-transport');
const { MESSAGE, LEVEL } = require('triple-beam');

const _noop = () => { };

// Ensure we have the correct winston here.
if (Number(winston.version.split('.')[0]) < 3) {
  throw new Error('Winston-syslog requires winston >= 3.0.0');
}

const levels = Object.keys({
  debug: 0,
  info: 1,
  notice: 2,
  warning: 3,
  warn: 3,
  error: 4,
  err: 4,
  crit: 5,
  alert: 6,
  emerg: 7
});

//
// ### function Syslog (options)
// #### @options {Object} Options for this instance.
// Constructor function for the Syslog Transport capable of sending
// RFC 3164 and RFC 5424 compliant messages.
//
class Syslog extends Transport {
  //
  // Expose the name of this Transport on the prototype
  //
  get name() {
    return 'syslog';
  }

  constructor(options = {}) {
    //
    // Inherit from `winston-transport`.
    //
    super(options);

    //
    // Setup connection state
    //
    this.connected = false;
    this.congested = false;
    this.retries = 0;
    this.queue = [];
    this.inFlight = 0;

    //
    // Merge the options for the target Syslog server.
    //
    this.setOptions(options);

    //
    // Setup our Syslog and network members for later use.
    //
    this.socket = null;
    var Producer = options.customProducer || glossy.Produce;
    this.producer = new Producer({
      type: this.type,
      appName: this.appName,
      pid: this.pid,
      facility: this.facility
    });
  }

  setOptions(options) {
    this.host = options.host || 'localhost';
    this.port = options.port || 514;
    this.path = options.path || null;
    this.protocol = options.protocol || 'udp4';
    this.protocolOptions = options.protocolOptions || {};
    this.endOfLine = options.eol;

    this.parseProtocol(this.protocol);
    // this options provides a secure socket for sending logs to the syslog server based on
    // TLSSocket options (https://nodejs.org/api/tls.html#class-tlstlssocket).
    // Therefore, TLSSocket options are configurable through the code.
    this.secureProtocol = options.secureProtocol;
    this.ciphers = options.ciphers;
    this.ecdhCurve = options.ecdhCurve;
    // end options for secure socket
    //
    // Merge the default message options.
    //
    this.localhost =
      typeof options.localhost !== 'undefined'
        ? options.localhost
        : 'localhost';
    this.type = options.type || 'BSD';
    this.facility = options.facility || 'local0';
    this.pid = options.pid || process.pid;
    this.appName = options.appName || options.app_name || process.title;
  }

  parseProtocol(protocol = this.protocol) {
    const parsedProtocol = utils.parseProtocol(protocol);

    this.protocolType = parsedProtocol.type;
    this.protocolFamily = parsedProtocol.family;
    this.isDgram = parsedProtocol.isDgram;

    if (this.protocolType === 'unix' && !this.path) {
      throw new Error('`options.path` is required on unix dgram sockets.');
    }
  }

  //
  // ### function chunkMessage (buffer, callback)
  // #### @buffer {Buffer} Syslog message buffer.
  // #### @callback {function} Continuation to respond to when complete.
  // Syslog messages sent over the UDP transport must be 64KB bytes or less. In
  // order to avoid silent failures messages should be chunked when the buffer
  // to write is larger than the maximum message size.
  //
  chunkMessage(buffer, callback = _noop) {
    if (!this.connected) {
      this.queue.push(buffer);
    } else {
      // Maximum payload for the UDP transport
      // 65535 − 8 bytes UDP header − 20 bytes IP header
      // https://tools.ietf.org/html/rfc5426#section-3.2
      // This makes sense on loopback interfaces with MTU = 65536
      // For non-loopback messages, it's impossible to know in advance
      // the MTU of each interface through which a packet might be routed
      // https://nodejs.org/api/dgram.html
      const MAX_UDP_PAYLOAD = 65507;
      let offset = 0;

      while (offset < buffer.length) {
        this.inFlight++;
        const length =
          offset + MAX_UDP_PAYLOAD > buffer.length
            ? buffer.length - offset
            : MAX_UDP_PAYLOAD;
        this._sendChunk(
          buffer,
          { offset: offset, length: length, port: this.port, host: this.host },
          callback
        );
        offset += length;
      }
    }
  }

  //
  // ### function _sendChunk (buffer, options, callback)
  // #### @buffer {Buffer} Syslog message buffer.
  // #### @options {object} Options for the message send method.
  // #### @callback {function} Continuation to respond to when complete.
  // Sends a single chunk from an oversize UDP buffer.
  //
  _sendChunk(buffer, options, callback) {
    try {
      this.socket.send(
        buffer,
        options.offset,
        options.length,
        options.port,
        options.host,
        callback
      );
    } catch (err) {
      return callback(err);
    }
  }

  //
  // ### function log (info, callback)
  // #### @info {object} All relevant log information
  // #### @callback {function} Continuation to respond to when complete.
  // Core logging method exposed to Winston. Logs the `msg` and optional
  // metadata, `meta`, to the specified `level`.
  //
  log(info, callback) {
    let level = info[LEVEL];
    if (!~levels.indexOf(level)) {
      return callback(
        new Error('Cannot log unknown syslog level: ' + info[LEVEL])
      );
    }
    level = level === 'warn' ? 'warning' : level;
    const output = info[MESSAGE];

    const syslogMsg = this.producer.produce({
      severity: level,
      host: this.localhost,
      date: new Date(),
      message: this.endOfLine ? output + this.endOfLine : output
    });

    //
    // Attempt to connect to the socket
    //
    this.connect(err => {
      if (err) {
        //
        // If there was an error enqueue the message
        //
        this.queue.push(syslogMsg);

        return callback();
      }

      //
      // On any error writing to the socket, enqueue the message
      //
      const onError = logErr => {
        if (logErr) {
          this.queue.push(syslogMsg);
          this.emit('error', logErr);
        }
        this.emit('logged', info);
        this.inFlight--;
      };

      const onCongestion = () => {
        onError(new Error('Congestion Error'));
      };

      const sendDgram = () => {
        const buffer = Buffer.from(syslogMsg);

        if (this.protocolType === 'udp') {
          this.chunkMessage(buffer, onError);
        } else if (this.protocol === 'unix') {
          this.inFlight++;
          this.socket.send(buffer, 0, buffer.length, this.path, onError);
        } else if (this.congested) {
          this.queue.push(syslogMsg);
        } else {
          this.socket.once('congestion', onCongestion);
          this.inFlight++;
          this.socket.send(buffer, e => {
            this.socket.removeListener('congestion', onCongestion);
            onError(e);
          });
        }
      };

      //
      // Write to the `tcp*`, `udp*`, or `unix` socket.
      //
      if (this.isDgram) {
        sendDgram();
      } else {
        this.socket.write(syslogMsg, 'utf8', onError);
      }

      callback(null, true);
    });
  }

  //
  // ### function close ()
  // Closes the socket used by this transport freeing the resource.
  //
  close() {
    const max = 6;
    let attempt = 0;

    const _close = () => {
      if (attempt >= max || (this.queue.length === 0 && this.inFlight <= 0)) {
        if (this.socket) {
          if (this.socket.destroy) {
            // https://nodejs.org/api/net.html#net_socket_destroy_exception
            this.socket.destroy();
          } else if (this.socket.close) {
            // https://nodejs.org/api/dgram.html#dgram_socket_close_callback
            // https://www.npmjs.com/package/unix-dgram#socketclose
            this.socket.close();
          }
        }

        this.emit('closed', this.socket);
      } else {
        attempt++;
        setTimeout(_close, 200 * attempt);
      }
    };
    _close();
  }

  connectDgram(callback) {
    if (this.protocol === 'unix-connect') {
      return this._unixDgramConnect(callback);
    } else if (this.protocol === 'unix') {
      this.socket = require('unix-dgram').createSocket('unix_dgram');
    } else if (!this.socket) {
      // UDP protocol
      const proto = this.protocol === 'udp' ? 'udp4' : this.protocol;
      // https://nodejs.org/api/all.html#dgram_class_dgram_socket
      this.socket = dgram.createSocket({
        type: proto
      })
        .on('listening', () => {
          this.connected = true;
          let msg = this.queue.shift();
          while (msg) {
            this.chunkMessage(msg);
            msg = this.queue.shift();
          }
        });
      this.socket.bind();
    }

    return callback(null);
  }
  //
  // ### function connect (callback)
  // #### @callback {function} Continuation to respond to when complete.
  // Connects to the remote syslog server using `dgram` or `net` depending
  // on the `protocol` for this instance.
  //
  connect(callback) {
    //
    // If the socket already exists then respond
    //
    if (this.socket) {
      return !this.socket.readyState ||
        this.socket.readyState === 'open' ||
        this.socket.connected
        ? callback(null)
        : callback(true);
    }

    //
    // Create the appropriate socket type.
    //
    if (this.isDgram) {
      return this.connectDgram(callback);
    }
    // Set Secure options for TLS
    const options = {
      secureProtocol: this.secureProtocol,
      ciphers: this.ciphers,
      ecdhCurve: this.ecdhCurve
    };

    this.socket = /^tls[4|6]?$/.test(this.protocol)
      ? new secNet.TLSSocket(this.socket, options) // Set options for TLSSocket
      : new net.Socket();
    this.socket.setKeepAlive(true);
    this.socket.setNoDelay();

    this.setupEvents();

    const connectConfig = Object.assign({}, this.protocolOptions, {
      host: this.host,
      port: this.port
    });

    if (this.protocolFamily) {
      connectConfig.family = this.protocolFamily;
    }

    this.socket.connect(connectConfig);

    //
    // Indicate to the callee that the socket is not ready. This
    // will enqueue the current message for later.
    //
    callback(true);
  }

  setupEvents() {
    const readyEvent = 'connect';
    //
    // On any error writing to the socket, emit the `logged` event
    // and the `error` event.
    //
    const onError = logErr => {
      if (logErr) {
        this.emit('error', logErr);
      }
      this.emit('logged');
      this.inFlight--;
    };

    //
    // Listen to the appropriate events on the socket that
    // was just created.
    //
    this.socket
      .on(readyEvent, () => {
        //
        // When the socket is ready, write the current queue
        // to it.
        //
        this.socket.write(this.queue.join(''), 'utf8', onError);

        this.emit('logged');
        this.queue = [];
        this.retries = 0;
        this.connected = true;
      })
      .on('error', function () {
        //
        // TODO: Pass this error back up
        //
      })
      .on('close', () => {
        //
        // Attempt to reconnect on lost connection(s), progressively
        // increasing the amount of time between each try.
        //
        const interval = Math.pow(2, this.retries);
        this.connected = false;

        setTimeout(() => {
          this.retries++;
          this.socket.connect(this.port, this.host);
        }, interval * 1000);
      })
      .on('timeout', () => {
        if (this.socket.destroy) {
          // https://nodejs.org/api/net.html#net_socket_settimeout_timeout_callback
          this.socket.destroy();
        } else if (this.socket.close) {
          // https://nodejs.org/api/dgram.html#dgram_socket_close_callback
          // https://www.npmjs.com/package/unix-dgram#socketclose
          this.socket.close();
        }
      });
  }

  _unixDgramConnect(callback) {
    const self = this;

    const flushQueue = () => {
      let sentMsgs = 0;
      this.queue.forEach(msg => {
        const buffer = Buffer.from(msg);

        if (!this.congested) {
          if (this.protocolType === 'udp') {
            this.chunkMessage(buffer, () => {
              ++sentMsgs;
            });
          } else {
            this.socket.send(buffer, function () {
              ++sentMsgs;
            });
          }
        }
      });

      this.queue.splice(0, sentMsgs);
    };

    this.socket = require('unix-dgram').createSocket('unix_dgram');
    this.socket.on('error', err => {
      this.emit('error', err);

      if (err.syscall === 'connect') {
        this.socket.close();
        this.socket = null;
        return callback(err);
      }
      if (err.syscall === 'send') {
        this.socket.close();
        this.socket = null;
      }
    });

    this.socket.on('connect', function () {
      this.on('congestion', () => {
        self.congested = true;
      });

      this.on('writable', () => {
        self.congested = false;
        flushQueue();
      });

      flushQueue();
      callback();
    });

    this.socket.connect(this.path);
  }
}

//
// Define a getter so that `winston.transports.Syslog`
// is available and thus backwards compatible.
//
winston.transports.Syslog = Syslog;

module.exports = {
  Syslog
};
